### PR TITLE
[Profiler] Make `RecordQueue` manage the lifetime of `PythonTracer`.

### PR DESCRIPTION
**Pull request by robieta**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #84276
* #84275
* #84274
* #84273
* #83965
* __->__ #83964
* #83895
* #84668
* #84667

`PythonTracer` holds a pointer to an owning `RecordQueue`, however that relationship is not enforced and it is possible to dangle that pointer if the ProfilerState owning the `RecordQueue` is destroyed without proper cleanup.

We currently use a singleton to enforce the requirement that only one python tracer is active at a time, however a better formulation is to simply enforce that with an atomic bool and manage object lifetime through composition. In this new architecture, `RecordQueue` explicitly holds a unique_ptr to the python tracer instance. That way if `~RecordQueue` is called it will call `~PythonTracer` which can then clean up any state. Overall it is just a simpler ownership model, and less prone to unexpected failures.

Differential Revision: [D38955616](https://our.internmc.facebook.com/intern/diff/D38955616/)

### CODE REVIEW

File: torch/csrc/autograd/profiler_python.cpp [hunk 1]
```diff
@@ -512,51 +520,50 @@ struct ThreadLocalResults {
...
  private:
-  PythonTracer();
-
   void recordPyCall(ThreadLocalResults& tls, PyFrameObject* frame);
   void recordCCall(
       ThreadLocalResults& tls,
       PyFrameObject* frame,
       PyObject* arg);
 
+  std::atomic<bool> active_lock_{false};
```
Comment (slgong-fb) [inline torch/csrc/autograd/profiler_python.cpp:545]: I like `std::atmoic` used here.

**Review by slgong-fb — APPROVED** [review]

File: torch/csrc/autograd/profiler_python.cpp (hunk 1 above)
Comment (robieta) [reply torch/csrc/autograd/profiler_python.cpp:545]: Thanks!

**Pull request closed without merging**